XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,304 @@
+
+                    if (updateDownloadState == DownloadState.Cancelled)
+                        throw new CancelledException();
+
+                    if (updateDownloadState == DownloadState.Error)
+                    {
+                        needToRetry = true;
+
+                        // this many errors so far - including this one
+                        errorCount++;
+
+                        // logging only, it will retry again.
+                        log.ErrorFormat(
+                            "Error while downloading from '{0}'. Number of errors so far (including this): {1}. Trying maximum {2} times.",
+                            address, errorCount, MAX_NUMBER_OF_TRIES);
+
+                        if (updateDownloadError == null)
+                            log.Error("An unknown error occurred.");
+                        else
+                            log.Error(updateDownloadError);
+                    }
+                } while (errorCount < MAX_NUMBER_OF_TRIES && needToRetry);
+            }
+            finally
+            {
+                //deregister download events
+                client.DownloadProgressChanged -= client_DownloadProgressChanged;
+                client.DownloadFileCompleted -= client_DownloadFileCompleted;
+
+                NetworkChange.NetworkAvailabilityChanged -= NetworkAvailabilityChanged;
+
+                client.Dispose();
+            }
+
+            //if this is still the case after having retried MAX_NUMBER_OF_TRIES number of times.
+            if (updateDownloadState == DownloadState.Error)
+            {
+                log.ErrorFormat("Giving up - Maximum number of retries ({0}) has been reached.", MAX_NUMBER_OF_TRIES);
+
+                MarkCompleted(updateDownloadError ?? new Exception(Messages.ERROR_UNKNOWN));
+            }
+
+        }
+
+        private void NetworkAvailabilityChanged(object sender, NetworkAvailabilityEventArgs e)
+        {
+            if (!e.IsAvailable && client != null && updateDownloadState == DownloadState.InProgress)
+            {
+                updateDownloadError = new WebException(Messages.NETWORK_CONNECTIVITY_ERROR);
+                updateDownloadState = DownloadState.Error;
+                client.CancelAsync();
+            }
+        }
+
+        protected override void Run()
+        {
+            if (downloadUpdate)
+            {
+                log.InfoFormat("Downloading update '{0}' (from '{1}') to '{2}'", updateName, address, outputPathAndFileName);
+                Description = string.Format(Messages.DOWNLOAD_AND_EXTRACT_ACTION_DOWNLOADING_DESC, updateName);
+                LogDescriptionChanges = false;
+                DownloadFile();
+                LogDescriptionChanges = true;
+
+                if (IsCompleted || Cancelled)
+                    return;
+
+                if (Cancelling)
+                    throw new CancelledException();
+            }
+            
+            ValidateMsi();

```

Comment:
Because we need error messages based on the checksum or the valid cert check on the method throws exceptions that then appear as error messages in the action progress dialog. Returning true false doesn't give enough information as to what went wrong. I could split the method but its highly possibly once  Matthews feed back is in that this check will change.

Reply:
ok then let's leave it. I agree that it's more important to give better messages to the user